### ISSUE TITLE
ShedLock is working with Neo4j Community, but not work with Neo4j Enterprise

### ISSUE THREAD

**Issue by tle130475c** [lukas-krecan/ShedLock#2099]
**Describe the bug**
ShedLock is working with Neo4j Community, **but not work with Neo4j Enterprise**
The version of Neo4j (both Community and Enterprise) is `5.22.0`, but even on Neo4j `4.4.16` the bug still there.
Full demo code: https://github.com/tle130475c/demo-scheduling-shedlock-neo4j

1. Which version do you use: `5.15.0`
2. Which Lock Provider: `shedlock-provider-neo4j`
3. ShedLock configuration

Spring Boot deployed to Kubernetes, 2 pods by default

File `EmployeeService.java`
```
package com.tle130475c.demo_scheduling_shedlock_neo4j.employee;

import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import lombok.extern.slf4j.Slf4j;
import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;

@Service
@Slf4j
public class EmployeeService {
    @Scheduled(cron = "*/2 * * * * *")
    @SchedulerLock(name = "EmployeeService_sendBirthdayGreetings", lockAtLeastFor = "PT1S")
    public void sendBirthdayGreetings() {
        log.info("Sending birthday greetings to employees");
    }
}
```
File `ScheduledConfig.java`
```
package com.tle130475c.demo_scheduling_shedlock_neo4j.config;

import org.neo4j.driver.Driver;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;

import net.javacrumbs.shedlock.provider.neo4j.Neo4jLockProvider;

@Configuration
public class ScheduledConfig {
    @Bean
    Neo4jLockProvider lockProvider(Driver driver) {
        return new Neo4jLockProvider(driver);
    }

    @Bean
    TaskScheduler taskScheduler() {
        ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();

        taskScheduler.setPoolSize(10);
        taskScheduler.setThreadNamePrefix("CN-Scheduled-");
        taskScheduler.initialize();

        return taskScheduler;
    }
}
```
File `DemoSchedulingShedlockNeo4jApplication`
```
package com.tle130475c.demo_scheduling_shedlock_neo4j;

import org.neo4j.cypherdsl.core.renderer.Configuration;
import org.neo4j.cypherdsl.core.renderer.Dialect;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.scheduling.annotation.EnableScheduling;

import net.javacrumbs.shedlock.spring.annotation.EnableSchedulerLock;

@SpringBootApplication
@EnableScheduling
@EnableSchedulerLock(defaultLockAtMostFor = "PT30S")
public class DemoSchedulingShedlockNeo4jApplication {
    public static void main(String[] args) {
        SpringApplication.run(DemoSchedulingShedlockNeo4jApplication.class, args);
    }

    @Bean
    Configuration cypherDslConfiguration() {
        return Configuration.newConfig()
                .withDialect(Dialect.NEO4J_5).build();
    }
}
```
File `application.yaml`
```
spring:
  application:
    name: demo-scheduling-shedlock-neo4j
  neo4j:
    uri: bolt://${NEO4J_SERVICE}:7687
    authentication:
      username: ${NEO4J_USER}
      password: ${NEO4J_PASSWORD}
logging:
  level:
    "[net.javacrumbs.shedlock]": DEBUG

```
4. ShedLock logs

**Expected behavior**
In Kubernetes, if the task is being executed in one pod,  it acquires a lock which prevents execution of the same task from another pods no matter the version of Neo4j is Community or Enterprise.
Below are the logs when work with Neo4j Community (expected behavior):

`Logs from Pod 1`
```
2024-08-20T02:47:54.018Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:47:54.028Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:47:55Z
2024-08-20T02:47:56.018Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:48:26Z
2024-08-20T02:47:56.019Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:47:56.031Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:47:57Z
2024-08-20T02:47:58.011Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Not executing 'EmployeeService_sendBirthdayGreetings'. It's locked.
2024-08-20T02:48:00.022Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Not executing 'EmployeeService_sendBirthdayGreetings'. It's locked.
2024-08-20T02:48:02.011Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:48:32Z
2024-08-20T02:48:02.011Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:48:02.017Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:48:03Z
```
`Logs from Pod 2`
```
2024-08-20T02:47:52.020Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:47:53Z
2024-08-20T02:47:54.024Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Not executing 'EmployeeService_sendBirthdayGreetings'. It's locked.
2024-08-20T02:47:56.022Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Not executing 'EmployeeService_sendBirthdayGreetings'. It's locked.
2024-08-20T02:47:58.009Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:48:28Z
2024-08-20T02:47:58.009Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:47:58.014Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:47:59Z
2024-08-20T02:48:00.018Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:48:30Z
2024-08-20T02:48:00.018Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:48:00.027Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:48:01Z
2024-08-20T02:48:02.015Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-4] n.j.s.core.DefaultLockingTaskExecutor    : Not executing 'EmployeeService_sendBirthdayGreetings'. It's locked.


```

**Actual behavior**
The scheduled is executed parallel in all pods, no lock happened.
Below are the logs when work with Neo4j Enterprise:
`Logs from Pod 1`
```
2024-08-20T02:58:36.053Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-9] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:37.001Z
2024-08-20T02:58:38.040Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-9] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:08.001Z
2024-08-20T02:58:38.041Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-9] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:38.049Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-9] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:39.001Z
2024-08-20T02:58:40.034Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-3] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:10.001Z
2024-08-20T02:58:40.034Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-3] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:40.047Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-3] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:41.001Z
2024-08-20T02:58:42.016Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:12Z
2024-08-20T02:58:42.016Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:42.024Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:43
```
`Logs from Pod 2`
```
2024-08-20T02:58:36.051Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-6] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:37.001Z
2024-08-20T02:58:38.034Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-5] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:08.001Z
2024-08-20T02:58:38.034Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-5] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:38.047Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-5] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:39.001Z
2024-08-20T02:58:40.041Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [CN-Scheduled-10] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:10.001Z
2024-08-20T02:58:40.042Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [CN-Scheduled-10] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:40.050Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [CN-Scheduled-10] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:41.001Z
2024-08-20T02:58:42.021Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-2] n.j.s.core.DefaultLockingTaskExecutor    : Locked 'EmployeeService_sendBirthdayGreetings', lock will be held at most until 2024-08-20T02:59:12Z
2024-08-20T02:58:42.021Z  INFO 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-2] c.t.d.employee.EmployeeService           : Sending birthday greetings to employees
2024-08-20T02:58:42.027Z DEBUG 1 --- [demo-scheduling-shedlock-neo4j] [ CN-Scheduled-2] n.j.s.core.DefaultLockingTaskExecutor    : Task finished, lock 'EmployeeService_sendBirthdayGreetings' will be released at 2024-08-20T02:58:43Z
```

**Comment by lukas-krecan** [comment]
Hi, thanks for reporting. If I get you right, if the tasks are executed only on one pod, it works as expected no matter the license? Then I am afraid I will not be able to help you ase the debugging is going to be really tricky. Morevore, I have no idea how the enterprise version differs from the community. Sorry

**Comment by lukas-krecan** [comment]
The tests are [failing](https://github.com/lukas-krecan/ShedLock/actions/runs/10477752985) with the enterprise version. Not sure if it's going to help with the debugging.

**Comment by tle130475c** [comment]
Hi Thank you. I have pulled Neo4j Team in, they are also investigating on this issue.

**Comment by tle130475c** [comment]
Hi, with the help from Neo4j Team, I believe that the reason why ShedLock only works on Community but not on Enterprise version of Neo4j is due to the default runtime.

The enterprise is using the [pipelined runtime](https://neo4j.com/docs/cypher-manual/current/planning-and-tuning/runtimes/concepts/#runtimes-pipelined-runtime) by default while the community is using the [slotted runtime](https://neo4j.com/docs/cypher-manual/current/planning-and-tuning/runtimes/concepts/#runtimes-slotted-runtime). Queries using pipelined runtime are, unlike those run in slotted runtime, not being executed one at a time. Rather, the pipelined runtime allows the physical operators to handle batches of between roughly 100 and 1000 transactions each (referred to as morsels), which are written into buffers containing data and tasks for a pipeline. A pipeline can, in turn, be defined as a sequence of operators which have been fused into one another so that they may be executed together in the same task by the runtime.

Running ShedLock with Enterprise version, two transactions can be committed together in a single pipeline while the Community has to do it one by one in the slotted pipeline. That's where the problem comes in.

Prepend the query inside `updateRecord` method  with  `CYPHER runtime = slotted` make ShedLock work as expected on both Community and Enterprise version of Neo4j (even when I set lockAtLeastFor = 1s or not set it in my system).

So for this problem I think we need to fix it with pipelined runtime in mind. The above solution is work, but I don't know if it's the best solution.

**Comment by lukas-krecan** [comment]
Thank you, I will try that

**Comment by lukas-krecan** [comment]
Fix released as 5.15.1
